drivers/flash/flash_npcx_fiu_nor.c

```diff
@@ -0,0 +1,625 @@
+		flash_npcx_uma_cmd_only(dev, SPI_NOR_CMD_WREN);
+		ret = flash_npcx_uma_write_by_addr(dev, SPI_NOR_CMD_PP, tx_buf,
+						   sz_write, addr);
+		if (ret != 0) {
+			break;
+		}
+
+		/* Wait for writing completed */
+		ret = flash_npcx_nor_wait_until_ready(dev);
+		if (ret != 0) {
+			break;
+		}
+
+		size -= sz_write;
+		tx_buf += sz_write;
+		addr += sz_write;
+
+		if (size > SPI_NOR_PAGE_SIZE) {
+			sz_write = SPI_NOR_PAGE_SIZE;
+		} else {
+			sz_write = size;
+		}
+	}
+
+	return ret;
+}
+
+static const struct flash_parameters *
+flash_npcx_nor_get_parameters(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return &flash_npcx_parameters;
+};
+
+#ifdef CONFIG_FLASH_EX_OP_ENABLED
+static int flash_npcx_nor_ex_exec_uma(const struct device *dev,
+				      const struct npcx_ex_ops_uma_in *op_in,
+				      const struct npcx_ex_ops_uma_out *op_out)
+{
+	int flag = 0;
+	struct npcx_uma_cfg cfg;
+
+	if (op_in == NULL) {
+		return -EINVAL;
+	}
+
+	/* Organize a UMA transaction */
+	cfg.opcode = op_in->opcode;
+	if (op_in->tx_count != 0) {
+		cfg.tx_buf = op_in->tx_buf;
+		cfg.tx_count = op_in->tx_count;
+		flag |= NPCX_UMA_ACCESS_WRITE;
+	}
+
+	if (op_in->addr_count != 0) {
+		cfg.addr.u32 = sys_cpu_to_be32(op_in->addr);
+		flag |= NPCX_UMA_ACCESS_ADDR;
+	}
+
+	if (op_out != NULL && op_in->rx_count != 0) {
+		cfg.rx_buf = op_out->rx_buf;
+		cfg.rx_count = op_in->rx_count;
+		flag |= NPCX_UMA_ACCESS_READ;
+	}
+
+	return flash_npcx_uma_transceive(dev, &cfg, flag);
+}
+
+static int flash_npcx_nor_ex_set_spi_spec(const struct device *dev,
+					  const struct npcx_ex_ops_qspi_oper_in *op_in)
+{
+	struct flash_npcx_nor_data *data = dev->data;
+
+	/* Cannot disable write protection of internal flash */
+	if ((data->operation & NPCX_EX_OP_INT_FLASH_WP) != 0) {
+		if ((op_in->mask & NPCX_EX_OP_INT_FLASH_WP) != 0 && !op_in->enable) {
+			return -EINVAL;
+		}
+	}
+
+	if (op_in->enable) {
+		data->operation |= op_in->mask;
+	} else {
+		data->operation &= ~op_in->mask;
+	}
+
+	return 0;
+}
+
+static int flash_npcx_nor_ex_get_spi_spec(const struct device *dev,
+					  struct npcx_ex_ops_qspi_oper_out *op_out)
+{
+	struct flash_npcx_nor_data *data = dev->data;
+
+	op_out->oper = data->operation;
+	return 0;
+}
+
+static int flash_npcx_nor_ex_op(const struct device *dev, uint16_t code,
+				const uintptr_t in, void *out)
+{
+#ifdef CONFIG_USERSPACE
+	bool syscall_trap = z_syscall_trap();
+#endif
+	int ret;
+
+	switch (code) {
+	case FLASH_NPCX_EX_OP_EXEC_UMA:
+	{
+		struct npcx_ex_ops_uma_in *op_in = (struct npcx_ex_ops_uma_in *)in;
+		struct npcx_ex_ops_uma_out *op_out = (struct npcx_ex_ops_uma_out *)out;
+#ifdef CONFIG_USERSPACE
+		struct npcx_ex_ops_uma_in in_copy;
+		struct npcx_ex_ops_uma_out out_copy;
+
+		if (syscall_trap) {
+			Z_OOPS(z_user_from_copy(&in_copy, op_in, sizeof(in_copy)));
+			op_in = &in_copy;
+			op_out = &out_copy;
+		}
+#endif
+
+		ret = flash_npcx_nor_ex_exec_uma(dev, op_in, op_out);
+#ifdef CONFIG_USERSPACE
+		if (syscall_trap) {
```

Comment:
Non-blocking suggestion.  Don't bother with the copy to the user data on a failure.
```suggestion
		if (ret == 0 && syscall_trap) {
```

Reply:
Done.